### PR TITLE
Added MNDWI code and Flake8 fixes

### PR DESCRIPTION
**Pull request by Oceancolour-RG**
Added mndwi.py so that mndwi h5 files can be created from the tesp workflow. Additional Flake8 fixes (from GADI) were performed.

### CODE REVIEW

**Comment by sixy6e** [comment]
Looks like you need to update black:

`black, version 20.8b1`

File: eugl/mndwi.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+            "description": desc,
+            "n_lines": nRows,
+            "n_samples": nCols,
+            "platform": platform_id,
+            "spatial_resolution": spatial_res,
+        }
+
+        # add attrs to dataset
+        for key in attrs:
+            mndwi_ds.attrs[key] = attrs[key]
```
Comment (sixy6e) [inline eugl/mndwi.py]: Would be ideal to include some additional attrs.  You can use this [function](https://github.com/GeoscienceAustralia/wagl/blob/develop/wagl/hdf5/__init__.py#L62) to attach them, which also allows to pass through any other attributes that you also want attached.  That will forgo the need to loop through the attributes yourself.

File: eugl/mndwi.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            products[0],
+            psplit(rgb_paths[i])[-1],
+        )
+        attrs["description"] = desc
+
+        # add attrs to dataset
+        for key in attrs:
+            rgb_ds.attrs[key] = attrs[key]
+
+        # For some reason I have to recreate the tiles. Attempts
```
Comment (sixy6e) [inline eugl/mndwi.py]: Delete the comment.

A generator is a one life object.  If you want to keep it to loop again later convert the generator to a list. e.g.:

```python
tiles = list(generate_tiles(...))
```

File: eugl/mndwi.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            mndwi_tile = compute_mndwi(green_im[tile], swir_im[tile])
+
+            # perform masking
+            mndwi_tile[~np.isfinite(mndwi_tile)] = nodata
+            mndwi_tile[mask[tile]] = nodata
+
+            mndwi_ds[tile] = mndwi_tile
+            mndwi_im[tile] = mndwi_tile
+
+        lowerVal, UpperVal = np.percentile(mndwi_im[~mask], (1, 99))
```
Comment (sixy6e) [inline eugl/mndwi.py]: Looks like you're now down a constrast stretch.  Can you split this into a separate function.

File: eugl/mndwi.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    c3 = 10.0
+    c4 = -0.015
+
+    scaled_im = c1 + c2 * np.arctan(c3 * (band_im / sfactor + c4))
+    scaled_im[scaled_im < 0] = 0
+    scaled_im[scaled_im > 1] = 1
+    return np.array(255 * scaled_im, order="C", dtype=np.uint8)
+
+
+def mndwi(wagl_h5_file, granule, out_fname):
```
Comment (sixy6e) [inline eugl/mndwi.py:240]: This function call is fairly lengthy.

Ideally split it into different components.

The actual mndwi calculation ideally should be its own function, that takes the required arrays, applies the function and returns an array.

File: eugl/mndwi.py (hunk 1 above)
Comment (sixy6e) [reply eugl/mndwi.py]: It's specifically for defining IMAGE class datasets.

File: eugl/mndwi.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    # get platform name
+    md = yaml.load(fid[granule + "/METADATA/CURRENT"][()], Loader=yaml.FullLoader)
+    platform_id = md["source_datasets"]["platform_id"]
+
+    # Get the 1st (lowerVal_arr) and 99th (upperVal_arr) percentikes
+    # for the each of the products. These arrays are very useful
+    # for creating MNDWI images, but this code will be removed once
+    # testing is complete.
+    lowerVal_arr = np.zeros([len(products)], order="C", dtype=np.float32)
+    upperVal_arr = np.zeros([len(products)], order="C", dtype=np.float32)
```
Comment (sixy6e) [inline eugl/mndwi.py]: Put this into a separate function call along with the other components related to the contrast enhancement.

File: eugl/mndwi.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    red_path, grn_path, blu_path = get_RGB_bands(platform_id, products[0], paths)
+    rgb_paths = [red_path, grn_path, blu_path]
+    rgb_names = ["red", "green", "blue"]
+
+    # create a group
+    rgb_grp = h5_fid.create_group("seadas_rgb")
+
+    # because no reprojection is necessary, we can load
+    # the red & blue bands as tiles to conserve memory
+    for i in range(0, len(rgb_names)):
```
Comment (sixy6e) [inline eugl/mndwi.py]: No need to include `0`.  Calling `range(n)` will always start at 0.

File: eugl/mndwi.py (hunk 3 above)
Comment (sixy6e) [reply eugl/mndwi.py]: Also, stick with `snake_case` for standard variable names.  We use `CamelCase` for defining classes.

**Review by sixy6e — CHANGES_REQUESTED** [review]
The function needs to be cut down a bit, and partitioned out into smaller function calls.

File: eugl/mndwi.py (hunk 1 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py]: What other additional attrs should I include?

File: eugl/mndwi.py (hunk 2 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py]: Ahhh thanks - fixed

File: eugl/mndwi.py (hunk 5 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py]: I've removed the lowerVal_arr and upperVal_arr as I can work around this when generating comparison figures

File: eugl/mndwi.py (hunk 4 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py:240]: `mndwi()` calls `compute_mndwi()` which computes the mndwi using the input green and swir arrays.

File: eugl/mndwi.py (hunk 6 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py]: fixed

File: eugl/mndwi.py (hunk 3 above)
Comment (Oceancolour-RG) [reply eugl/mndwi.py]: Thanks I've made the contrast stretch a separate function and used standard variable name conventions

**Review by sixy6e — APPROVED** [review]
Nice, a lot cleaner.

**Pull request merged**